lib/compat.php

```diff
@@ -304,6 +304,253 @@ function gutenberg_warn_classic_about_blocks() {
 	<?php
 }
 
+/**
+ * Add 'actual_size' to the prepared attachment data for the Media Library.
+ *
+ * Needed as `wp_prepare_attachment_for_js()` (for the media modal) constrains the image sizes
+ * to the theme's `$content_width`.
+ *
+ * @param array   $response   Array of prepared attachment data.
+ * @param WP_Post $attachment Attachment object.
+ * @param array   $meta       Array of attachment meta data.
+ * @return array Array of prepared attachment data.
+ */
+function gutenberg_prepare_attachment_for_js( $response, $attachment, $meta ) {
+	if ( ! empty( $response['type'] ) && 'image' === $response['type'] && ! empty( $response['sizes'] ) ) {
+		foreach ( $response['sizes'] as $size_name => $size ) {
+			if ( array_key_exists( $size_name, $meta['sizes'] ) ) {
+				$response['sizes'][ $size_name ]['actual_size'] = array(
+					'width'  => (int) $meta['sizes'][ $size_name ]['width'],
+					'height' => (int) $meta['sizes'][ $size_name ]['height'],
+				);
+			}
+		}
+	}
+
+	return $response;
+}
+add_filter( 'wp_prepare_attachment_for_js', 'gutenberg_prepare_attachment_for_js', 10, 3 );
+
+/**
+ * Warm the object cache with post and meta information for all found
+ * image blocks to avoid making individual database calls
+ * (similarly to `wp_make_content_images_responsive()`).
+ *
+ * @access private
+ * @since 4.x.x
+ *
+ * @param string $content The post content.
+ * @return string $content Unchanged post content.
+ */
+function _gutenberg_cache_images_meta( $content ) {
+	// Need to find all image blocks and get the attachment IDs from them BEFORE the parser is run
+	// so we can get the image attachments meta all at once from the DB.
+	if ( preg_match_all( '/^<!-- wp:image {.*"id":(\d+),.*} -->$/m', $content, $matches ) ) {
```

Comment:
Two issues with the regular expression:

- Lazy `.*` will potentially capture across multiple image blocks in the same line
- There's an enforced comma after `id` which assumes that `id` can't be the last property of the JSON (which we shouldn't assume)

https://regex101.com/r/SX8DA0/1

Reply:
> Greedy .* will potentially capture across multiple image blocks in the same line

Yeah, this assumes that the "block boundaries" are formatted properly, i.e. each is on a new line. Then there will only be one image block with one ID. That makes it somewhat fragile, but can't think of anything better that's fast enough... Still, planning to come back to this later (if possible).